speech_recognition/rnnt/pytorch/rnn.py

```diff
@@ -96,12 +96,15 @@ def __init__(self, factor):
 
     def forward(self, x, x_lens):
         # T, B, U
-        seq = [x]
-        for i in range(1, self.factor):
-            # This doesn't seem to make much sense...
-            tmp = torch.zeros_like(x)
-            tmp[:-i, :, :] = x[i:, :, :]
-            seq.append(tmp)
-        x_lens = torch.ceil(x_lens.float() / self.factor).int()
-        # Gross, this is horrible. What a waste of memory...
-        return torch.cat(seq, dim=2)[::self.factor, :, :], x_lens
+        r = torch.transpose(x, 0, 1)
+        s = r.shape
+        zeros = torch.zeros(
+            s[0], (-s[1]) % self.factor, s[2], dtype=r.dtype, device=r.device)
+        r = torch.cat([r, zeros], 1)
+        s = r.shape
+        rs = [s[0], s[1] // self.factor, s[2] * self.factor]
+        r = torch.reshape(r, rs)
+        numer = x_lens + (self.factor * 2 - 1)
+        x_lens = torch.div(numer, self.factor, rounding_mode="trunc") - 1
```

Comment:
It's late, so I'm just going to trust that this calculation of x_lens is correct.

Reply:
While double-checking my math, I realized that it's equivalent to the old calculation:

&lceil;x_lens/factor&rceil; = &lfloor;(x_lens+factor-1)/factor&rfloor; = &lfloor;(x_lens+factor-1)/factor + factor/factor&rfloor; - 1 = &lfloor;(x_lens+2*factor-1)/factor&rfloor; - 1

because (&lceil;a/b&rceil; = &lfloor;(a+b-1)/b&rfloor;) and  (&lfloor;c&rfloor; = &lfloor;c+1&rfloor; - 1).

In hindsight, I guess it's obvious that I didn't need to change the x_lens calculations, and I'll stick with the old calculation because it's simpler.